### PR TITLE
Allow to set K8S client QPS and burst values and bump default settings to 50/100

### PR DESCRIPTION
**Pull request by mwielgus**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Currently there is no easy easy way to increase the default QPS/Burst settings of 20/30 (imposed by controller-runtime). This PR adds QPS/Burst settings to the config file and bumps the default values to 25/35. Larger deployments may require increasing these values even further.

Before:
```
Not printed but would be 20/30
```
After:
```
2023-01-12T13:39:05.056718+01:00	LEVEL(-2)	setup	kueue/main.go:101	K8S Client	{"qps": 25, "burst": 35}
```

### CODE REVIEW

File: apis/config/v1alpha2/configuration_types.go [hunk 1]
```diff
@@ -51,6 +51,10 @@ type Configuration struct {
 	// This is achieved by blocking the start of new jobs until the previously
 	// started job has all pods running (ready).
 	WaitForPodsReady *WaitForPodsReady `json:"waitForPodsReady,omitempty"`
+
+	// Client provides additional configuration options for Kubernetes
+	// API server client.
+	Client *Client `json:"client,omitempty"`
```
Comment (alculquicondor) [inline apis/config/v1alpha2/configuration_types.go:57]: ClientConnection, to match kube-scheduler

File: config/components/manager/controller_manager_config.yaml [hunk 2]
```diff
@@ -16,6 +16,9 @@ controller:
     ClusterQueue.kueue.x-k8s.io: 3
     ResourceFlavor.kueue.x-k8s.io: 3
     Workload.kueue.x-k8s.io: 3
+client:
+  qps: 25
+  burst: 35
```
Comment (alculquicondor) [inline config/components/manager/controller_manager_config.yaml:21]: A bit too shy? :)

I think we can match kube-scheduler's default of qps=50, burst=100

https://github.com/kubernetes/kubernetes/blob/4802d7bb62c2623be8e4f940f6b5c1fcddd6c744/pkg/scheduler/apis/config/v1/defaults.go#L145

File: main.go [hunk 3]
```diff
@@ -92,7 +92,13 @@ func main() {
 	if kubeConfig.UserAgent == "" {
 		kubeConfig.UserAgent = useragent.Default()
 	}
-
+	if cfg.Client != nil && cfg.Client.QPS != nil {
```
Comment (alculquicondor) [inline main.go:95]: we set these defaults in apis/config/v1alpha2/defaults.go

File: apis/config/v1alpha2/configuration_types.go (hunk 1 above)
Comment (mwielgus) [reply apis/config/v1alpha2/configuration_types.go:57]: ComponentConfig has couple more fields that would require a slightly bigger refactoring to get them supported (like dropping kubefile flag etc). Not a big deal but not sure if it is actually needed.

File: main.go (hunk 3 above)
Comment (mwielgus) [reply main.go:95]: Do you mean that if QPS is not set we should default it to controller-runtime default of 20?

File: apis/config/v1alpha2/configuration_types.go (hunk 1 above)
Comment (alculquicondor) [reply apis/config/v1alpha2/configuration_types.go:57]: As long as we stay consistent, we can swap the types in the future.

File: main.go (hunk 3 above)
Comment (alculquicondor) [reply main.go:95]: no, I mean that this is not the correct place in the codebase to put the defaults. It should be in the apis/config/v1alpha2/defaults.go file.

File: apis/config/v1alpha2/configuration_types.go (hunk 1 above)
Comment (mwielgus) [reply apis/config/v1alpha2/configuration_types.go]: Done.

File: Makefile [hunk 4]
```diff
@@ -232,6 +232,7 @@ artifacts: kustomize
 	if [ -d artifacts ]; then rm -rf artifacts; fi
 	mkdir -p artifacts
 	$(KUSTOMIZE) build config/default -o artifacts/manifests.yaml
+	$(KUSTOMIZE) build config/dev -o artifacts/manifests-dev.yaml
```
Comment (alculquicondor) [inline Makefile:235]: there seems to be something wrong with the rebase

Comment (mwielgus) [reply Makefile]: Fixed.

File: config/components/manager/controller_manager_config.yaml (hunk 2 above)
Comment (mwielgus) [reply config/components/manager/controller_manager_config.yaml]: Done.

File: apis/config/v1alpha2/defaults.go [hunk 5]
```diff
@@ -29,6 +29,8 @@ const (
 	DefaultHealthProbeBindAddress = ":8081"
 	DefaultMetricsBindAddress     = ":8080"
 	DefaultLeaderElectionID       = "c1f6bfd2.kueue.x-k8s.io"
+	DefaultClientConnectionQPS    = 20.0
+	DefaultClientConnectionBurst  = 30
```
Comment (alculquicondor) [inline apis/config/v1alpha2/defaults.go:33]: I would leave these ones unexported

File: apis/config/v1alpha2/configuration_types.go [hunk 6]
```diff
@@ -77,3 +81,12 @@ type InternalCertManagement struct {
 	// Defaults to kueue-webhook-server-cert.
 	WebhookSecretName *string `json:"webhookSecretName,omitempty"`
 }
+
+type ClientConnection struct {
+	// QPS controls the number of queries per second allowed for K8S api server
+	// connection.
+	QPS *float32 `json:"qps,omitempty"`
```
Comment (alculquicondor) [inline apis/config/v1alpha2/configuration_types.go:88]: https://github.com/kubernetes/kubernetes/blob/c0c386b9c9c0eb798e8ec8d37fb0b2dd6e4554d7/staging/src/k8s.io/component-base/config/v1alpha1/types.go#L79
```suggestion
	QPS float32 `json:"qps,omitempty"`
```

`int32` below

File: apis/config/v1alpha2/defaults_test.go [hunk 7]
```diff
@@ -199,6 +211,47 @@ func TestSetDefaults_Configuration(t *testing.T) {
...
+				InternalCertManagement: &InternalCertManagement{
+					Enable: pointer.Bool(false),
+				},
+				ClientConnection: &ClientConnection{
+					QPS:   pointer.Float32(123.0),
+					Burst: pointer.Int(456),
+				},
+			},
+		},
+		"should default empty custom ClientConnection": {
```
[the remark this answers is not in the document]
Comment (mwielgus) [reply apis/config/v1alpha2/defaults_test.go:240]: Not, here empty non nil clientConnection is provided.

File: apis/config/v1alpha2/defaults_test.go [hunk 8]
```diff
@@ -184,6 +192,10 @@ func TestSetDefaults_Configuration(t *testing.T) {
 					WebhookServiceName: pointer.String(DefaultWebhookServiceName),
 					WebhookSecretName:  pointer.String(DefaultWebhookSecretName),
 				},
+				ClientConnection: &ClientConnection{
```
[the remark this answers is not in the document]
Comment (mwielgus) [reply apis/config/v1alpha2/defaults_test.go:195]: Done.

File: apis/config/v1alpha2/configuration_types.go (hunk 6 above)
Comment (mwielgus) [reply apis/config/v1alpha2/configuration_types.go:88]: Everywhere else in this config we have pointers. I don't see a reason to break convention to potentially reuse a struct with 3 additional not-so-needed fields.

File: apis/config/v1alpha2/defaults.go (hunk 5 above)
Comment (mwielgus) [reply apis/config/v1alpha2/defaults.go:33]: they are needed in main_test.go

File: apis/config/v1alpha2/defaults_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (alculquicondor) [reply apis/config/v1alpha2/defaults_test.go:237]: Ah right

File: apis/config/v1alpha2/configuration_types.go (hunk 6 above)
Comment (alculquicondor) [reply apis/config/v1alpha2/configuration_types.go:88]: Sounds good, just move from `*int` to `*int32`

Comment (mwielgus) [reply apis/config/v1alpha2/configuration_types.go:88]: Done.

**Comment by alculquicondor** [comment]
uhm... again 1.23 being too slow to start the kueue controller manager :\

/retest

/lgtm
/approve

**Pull request merged**